### PR TITLE
Moved array_free after logprintf

### PR DESCRIPTION
**Pull request by woutput**
This fixed the segmentation fault described in [this thread] (http://forum.pilight.org/Thread-Bug-report-segmentation-fault-using-rule-with-AFTER)

### CODE REVIEW

File: libs/pilight/events/actions/switch.c [hunk 1]
```diff
@@ -174,9 +174,9 @@ static int checkArguments(struct rules_actions_t *obj) {
 								break;
 							}
 						}
-						array_free(&array, l);
 						if(match == 0) {
 							logprintf(LOG_ERR, "switch action \"%s\" is not a valid unit", array[1]);
+							array_free(&array, l);
 							return -1;
 						}
```
Comment (CurlyMoo) [inline libs/pilight/events/actions/switch.c]: It should be on line 181 as well or we get a memory leak when there are no errors.

**Comment by woutput** [comment]
Very true, sorry and well spotted. I will add another commit.

**Comment by woutput** [comment]
Now checked the entire file:
* Free should be done before every return
* Free should only be done when there is something to free

Currently compiling before test

**Comment by woutput** [comment]
Seems to be working fine.
Please review again and merge if OK.

**Comment by woutput** [comment]
Oops, forgot: if you think it's OK, I will merge into 1 commit first :smile:

**Comment by pilino1234** [comment]
Does this also affect other actions such as dimming?
For [https://github.com/pilight/pilight/blob/development/libs/pilight/events/actions/dim.c#L169](example).

**Comment by woutput** [comment]
Very well spotted @pilino1234 
Shall I combine the fixes in this pull request?

**Comment by pilino1234** [comment]
I'd say yes, rather than having several commits to do the same thing across different files.

**Comment by woutput** [comment]
I agree, I will search, fix & commit to this branch.

**Comment by pilino1234** [comment]
Can you test it to see if the actions work properly afterwards and valgrind is happy?

**Comment by woutput** [comment]
I don't know the tool valgrind, but apparently it hasn't spotted these errors.
Maybe somebody else (you?) could run valgrind after my changes?

**Comment by pilino1234** [comment]
Just use:
´´´valgrind --leak-check=full --show-leak-kinds=all ./pilight-daemon -D´´´

**Comment by woutput** [comment]
OK, I will do that :smile:

**Comment by woutput** [comment]
OK, I think that's it.
Will try to compile, run valgrind and test.

**Comment by woutput** [comment]
See [this post] (http://forum.pilight.org/Thread-Bug-report-segmentation-fault-using-rule-with-AFTER?pid=16127#pid16127)

**Comment by woutput** [comment]
Tests:
Before fix: 3 x segmentation fault
After fix: 3 x nice error message in /var/log/pilight.err

**Comment by woutput** [comment]
Valgrind:
Running command `valgrind --leak-check=full pilight-daemon -D` finds 1 error:
```
[Nov 22 11:51:55:626586] pilight-daemon: DEBUG: garbage collected event action library
[Nov 22 11:51:55:644058] pilight-daemon: DEBUG: garbage collected event function library
[Nov 22 11:51:55:650776] pilight-daemon: DEBUG: garbage collected events library
==8872== Syscall param access(pathname) points to unaddressable byte(s)
==8872==    at 0x4A6139C: access (syscall-template.S:82)
==8872==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==8872== 
[Nov 22 11:51:55:698595] pilight-daemon: DEBUG: garbage collected fcache library
[Nov 22 11:51:55:711376] pilight-daemon: DEBUG: garbage collected sha256cache library
[Nov 22 11:51:55:718012] pilight-daemon: DEBUG: garbage collected webserver library
```

**Comment by CurlyMoo** [comment]
Not an issue.

**Comment by pilino1234** [comment]
Just use Ctrl+C to terminate the pilight daemon after triggering those rules a bit. Then check the heap summary it gives you after pilight has stopped.

**Comment by woutput** [comment]
>Just use Ctrl+C to terminate the pilight daemon after triggering those rules a bit.

That didn't work since valgrind kills pilight: [valgrind_output.txt](https://github.com/pilight/pilight/files/41035/valgrind_output.txt)
This is part of the output:
```
==9028== 34 (16 direct, 18 indirect) bytes in 2 blocks are definitely lost in loss record 1,468 of 2,209
==9028==    at 0x4835A7C: realloc (vg_replace_malloc.c:632)
==9028==    by 0x39F9B: explode (common.c:138)
==9028==    by 0x7A2FF: checkArguments (switch.c:165)
==9028==    by 0x6C307: event_parse_action (events.c:1446)
==9028==    by 0x6D003: event_parse_rule (events.c:1688)
==9028==    by 0x65107: rules_parse (rules.c:104)
==9028==    by 0x31653: config_parse (config.c:119)
==9028==    by 0x31A1F: config_read (config.c:206)
==9028==    by 0x20EEB: start_pilight (daemon.c:2307)
==9028==    by 0x21CB3: main (daemon.c:2799)
```
which is about the command `explode` that is called in `checkArguments` in switch.c line 165 which I think is not my fault.

Any advice?

**Comment by woutput** [comment]
I propose to just continue (without the full confirmation of valgrind).
What do you think?

**Comment by CurlyMoo** [comment]
Will merge soon

**Comment by woutput** [comment]
Kind reminder @CurlyMoo :smiley:

**Pull request merged**